### PR TITLE
Azure Key Vault Health check package and Unit tests project

### PR DESCRIPTION
**Pull request by CarlosLanderas**
Azure Key Vault health check that allows to configure a secret collection to be verified against the Azure Service

Usage:

Default usage with Managed Service Identity:

```csharp
services.AddHealthChecks()
               .AddAzureKeyVault(setup =>
               {
                   setup
                   .UseKeyVaultUrl("https://[vaultname].vault.azure.net/")                  
                   .AddSecret("surname");
               });

```

Usage usage client id and client secret:

```csharp
  services.AddHealthChecks()
               .AddAzureKeyVault(setup =>
               {
                   setup
                   .UseKeyVaultUrl("https://[vaultname].vault.azure.net/")    
                   .UseClientSecrets("clientid", "clientsecret")
                   .AddSecret("surname");
               });
```

Sample error for a not found secret using UI.Client response writer

```json
{
    "status": "Unhealthy",
    "totalDuration": "00:00:01.0434203",
    "entries": {
        "azurekeyvault": {
            "data": {},
            "description": "surname secret error - Operation returned an invalid status code 'NotFound'",
            "duration": "00:00:01.0403080",
            "exception": "surname secret error - Operation returned an invalid status code 'NotFound'",
            "status": "Unhealthy"
        }
    }
}
```

### CODE REVIEW

File: build/dependencies.props [hunk 1]
```diff
@@ -77,11 +80,12 @@
     <HealthCheckDynamoDb>2.2.0</HealthCheckDynamoDb>
     <HealthCheckDocumentDb>2.2.0</HealthCheckDocumentDb>
     <HealthCheckAzureStorage>2.2.0</HealthCheckAzureStorage>
-    <HealthCheckAzureServiceBus>2.2.0</HealthCheckAzureServiceBus>
+    <HealthCheckAzureServiceBus>2.2.1</HealthCheckAzureServiceBus>
```
Comment (unaizorrilla) [inline build/dependencies.props]: why a new version on service bus

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace HealthChecks.AzureKeyVault
+{
+    public class AzureKeyVaultHealthCheck : IHealthCheck
+    {
+        private readonly AzureKeyVaultOptions _keyVaultOptions;
+
+        public AzureKeyVaultHealthCheck(AzureKeyVaultOptions keyVaultOptions)
+        {
+            if (string.IsNullOrEmpty(keyVaultOptions.KeyVaultUrlBase)) throw new ArgumentNullException(keyVaultOptions.KeyVaultUrlBase);
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs]: please add backets

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            if (string.IsNullOrEmpty(keyVaultOptions.ClientId)) throw new ArgumentNullException(keyVaultOptions.ClientId);
+            if (string.IsNullOrEmpty(keyVaultOptions.ClientSecret)) throw new ArgumentNullException(keyVaultOptions.ClientSecret);
+                
+            _keyVaultOptions = keyVaultOptions;
+        }
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            string currentSecret = string.Empty;
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs]: var

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            {
+                var client = new KeyVaultClient(GetToken);
+                foreach (var secretIdentifier in _keyVaultOptions.Secrets)
+                {
+                    currentSecret = secretIdentifier;
+                    await client.GetSecretAsync(_keyVaultOptions.KeyVaultUrlBase, secretIdentifier, cancellationToken);
+                }
+
+                return HealthCheckResult.Healthy();
+
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs]: remove blank lines

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            catch (Exception ex)
+            {
+                var secretException = new Exception($"{currentSecret} secret error - {ex.Message}", ex);
+                return new HealthCheckResult(context.Registration.FailureStatus, exception: secretException);
+            }
+        }
+
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs]: blank lines

File: src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        public async Task<string> GetToken(string authority, string resource, string scope)
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/AzureKeyVaultHealthCheck.cs]: what about Azure MSI? For Azure hosted applications KeyVaultClient can be created with AzureServiceTokenProvider and you don't need get token or specify clientId and client secret. **Add support for Azure MSI**

var azureServiceTokenProvider = new AzureServiceTokenProvider();
var keyVaultClient = new KeyVaultClient(
                            new KeyVaultClient.AuthenticationCallback(
                                azureServiceTokenProvider.KeyVaultTokenCallback));

File: src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+        /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+        /// </param>
+        /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+        /// <returns>The <see cref="IHealthChecksBuilder"/>.</returns></param>
+        public static IHealthChecksBuilder AddAzureKeyVault(this IHealthChecksBuilder builder, Action<AzureKeyVaultOptions> setup,
+            string name = default, HealthStatus? failureStatus = default, IEnumerable<string> tags = default)
+        {
+            var azureKeyVaultOptions = new AzureKeyVaultOptions();
+
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureKeyVault/DependencyInjection/AzureKeyVaultHealthChecksBuilderExtensions.cs]: blank line

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs [hunk 8]
```diff
@@ -13,8 +13,11 @@ public class AzureEventHubHealthCheck
         private readonly string _eventHubName;
         public AzureEventHubHealthCheck(string connectionString, string eventHubName)
         {
-            _connectionString = connectionString ?? throw new ArgumentNullException(nameof(connectionString));
-            _eventHubName = eventHubName ?? throw new ArgumentNullException(nameof(eventHubName));
```
Comment (unaizorrilla) [inline src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs]: probably mix features on different package is not a good idea for tracking project. Can you split in two different pull requests

File: test/UnitTests/DependencyInjection/HealthChecks.AzureKeyVault/AzureKeyVaultUnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azurekeyvault");
+            check.GetType().Should().Be(typeof(AzureKeyVaultHealthCheck));
+
```
Comment (unaizorrilla) [inline test/UnitTests/DependencyInjection/HealthChecks.AzureKeyVault/AzureKeyVaultUnitTests.cs]: blank lines

File: test/UnitTests/DependencyInjection/HeathChecks.AzureServiceBus/AzureServiceBusQueueUnitTests.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("azurequeue");
+            check.GetType().Should().Be(typeof(AzureServiceBusQueueHealthCheck));
+
```
Comment (unaizorrilla) [inline test/UnitTests/DependencyInjection/HeathChecks.AzureServiceBus/AzureServiceBusQueueUnitTests.cs]: blank lines

File: test/UnitTests/DependencyInjection/HeathChecks.AzureServiceBus/AzureServiceBusQueueUnitTests.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            registration.Name.Should().Be("azureservicebusqueuecheck");
+            check.GetType().Should().Be(typeof(AzureServiceBusQueueHealthCheck));
+        }
+
+        [Fact]
+        public void fail_when_no_health_check_configuration_provided()
+        {
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddAzureServiceBusQueue("", "");
```
Comment (unaizorrilla) [inline test/UnitTests/DependencyInjection/HeathChecks.AzureServiceBus/AzureServiceBusQueueUnitTests.cs]: use string.empty

**Comment by unaizorrilla** [comment]
Hi Carlos,
Thanks for summitting this PR, I review this and you need to fix some minor issues. Thanks for starting the new tests project for check registered healthchecks.

Some items to review:

- Please add support for Azure MSI into the KeyVaultPackage
-  Review minor issues  (backets and blank lines)
-  Split into two PR the fix for Servicebus HealthChecks preconditions changes

File: src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs (hunk 8 above)
Comment (CarlosLanderas) [reply src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs]: New tests can't pass without this changes to the package. :/

**Comment by omerlh** [comment]
What about checking also keys that should exist?

**Comment by CarlosLanderas** [comment]
That is on the PR as well. Using the AddSecret fluent api

**Comment by omerlh** [comment]
I was talking about KeyVault encryption keys - it's a different feature of KeyVault. I looked on the code, look like it's not checked...

**Comment by CarlosLanderas** [comment]
That's not added. The idea was checking the existing secrets that the application might need. Feel free to add it if you consider :)

**Comment by omerlh** [comment]
Can I add it to the existing PR? Or on a different one?

**Comment by CarlosLanderas** [comment]
You can use the existing one

**Comment by omerlh** [comment]
How can I build the project on mac? Should I set env var like `NetStandardTargetVersion`?

**Comment by unaizorrilla** [comment]
Hi @omerlh 

Right now on master, you can create a new PR for new features if you want!

**Comment by unaizorrilla** [comment]
Hi @CarlosLanderas 

Thanks for this, PR is merged on master with some minor details
